vibes-tools/tools/vibes-opt/lib/runner.ml

```diff
@@ -1,59 +1,73 @@
 open Core
+open Bap.Std
+open Bap_core_theory
 
-module T = Bap_core_theory.Theory
-module Log = Vibes_log_lib.Stream
-module Err = Vibes_error_lib.Std
-module Utils = Vibes_utils_lib
-module Serializers = Vibes_serializers_lib
-module Patch_info = Vibes_patch_info_lib.Types
-module Function_info = Vibes_function_info_lib.Types
+module T = Theory
+module Log = Vibes_log.Stream
+module Utils = Vibes_utils
+module Serializers = Vibes_serializers
+module Patch_info = Vibes_patch_info.Types
+module Function_info = Vibes_function_info.Types
+module Bir_helpers = Vibes_bir.Helpers
 
-open Vibes_error_lib.Let
+open KB.Syntax
 
-let log_bir bir =
-  Log.send (Format.asprintf "BIR:\n%a" Bap.Std.Blk.pp bir)
+let deserialize_and_opt
+    (sexp : Sexp.t)
+    ~(target : T.target)
+    ~(language : T.language)
+    ~(patch_info : Patch_info.t)
+    ~(func_info : Function_info.t) : sub term KB.t =
+  let* sub = Serializers.Bir.deserialize sexp in
+  Log.send "Deserialized BIR:\n\n%a" Sub.pp sub;
+  Bir_passes.run sub ~target ~language ~patch_info ~func_info
 
-let run (target : string) (language : string) (patch_info_filepath : string)
-    (bir_filepath : string) (func_info_filepath : string)
-    (outfile : string) : (unit, Err.t) result =
-  let msg = Format.sprintf
-    "Vibes_opt_lib.Runner.run '%s' '%s' '%s' '%s' '%s' '%s'"
-    target language patch_info_filepath
-    bir_filepath func_info_filepath outfile
-  in
-  Log.send msg;
+(* Try to deserialize and optimize the BIR program while
+   preserving the toplevel state. *)
+let try_deserialize_and_opt
+    (bir_sexp : Sexp.t)
+    ~(target : T.target)
+    ~(language : T.language)
+    ~(patch_info : Patch_info.t)
+    ~(func_info : Function_info.t) : (sub term, KB.conflict) result =
+  try
+    let result = Toplevel.var "vibes-opt" in
+    Toplevel.put result @@ deserialize_and_opt bir_sexp
+      ~target ~language ~patch_info ~func_info;
+    Result.return @@ Toplevel.get result
+  with Toplevel.Conflict err -> Error err
 
+let run
+    ~(target : string)
+    ~(language : string)
+    ~(patch_info_filepath : string)
+    ~(bir_filepath : string)
+    ~(func_info_filepath : string)
+    ~(bir_outfile : string) : (unit, KB.conflict) result =
+  let (let*) x f = Result.bind x ~f in
```

Comment:
Don't we already get this `let*` from `open KB.Syntax` above? Or did I forget what was in `KB.Syntax`?

Reply:
This is for the result monad, not the KB monad.